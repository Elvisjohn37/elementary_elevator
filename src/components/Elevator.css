.Elevator {
  height: fit-content;
  margin: 10px;
  padding: 10px;
  background-color: rgba(0, 0, 0, 0.082);
}

.Elevator-move-requests {
  position: absolute;
  top: 50px;
}

.Elevator-move-requests ul {
  display: contents;
}

.Elevator-move-requests ul li {
  list-style: none;
}

.Elevator-car {
  border: 1px solid #000;
  margin: 10px;
  padding: 10px;
  min-width: 200px;
  min-height: 400px;
}
