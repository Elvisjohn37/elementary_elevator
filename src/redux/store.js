import { configureStore } from "@reduxjs/toolkit";
import elevatorSlice from "./../components/elevator/slice";

const store = configureStore({
  reducer: {
    elevator: elevatorSlice,
  },
});

export default store;
