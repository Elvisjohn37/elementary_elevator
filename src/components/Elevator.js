import React from "react";
import "./Elevator.css";
import { useDispatch } from "react-redux";
import {
  moveUp,
  moveDown,
  changeStatus,
  changeMove,
  debarkPassengers,
  setInsidePassengers,
  requestIn,
} from "./elevator/slice";
import FloorForm from "./elevator/FloorForm";
import {
  getPassengers,
  getCurrentFloorPassengers,
  getPassengersToDebark,
  getLastFloor,
  getReverseMove,
  getUnsettledFloorTo,
  getPassengerMove,
  generateRandomFloor,
} from "./elevator/helper";

const Elevator = ({
  currentFloor,
  name,
  status,
  id,
  maxFloor,
  passengers,
  move,
}) => {
  const [isShowFloorForm, setIsShowFloorForm] = React.useState(false);
  const dispatch = useDispatch();
  const openFloorTime = 10000;
  const moveFloorTime = 10000;
  const [isFloorFrom, setIsFloorFrom] = React.useState(true);
  const [userMove, setUserMove] = React.useState(move);

  React.useEffect(() => {
    let elevatorTimeout;
    let elevatorTime = 0;
    let lastFloor;
    // Elevator's priority is to pick up the passengers on the last floor that hs passengers
    let isPriorityFrom = false;
    // Get all passengers that has same move with the current move of the elevator where the passengers floor is greater than or equal to the elevator's current floor.
    let currentMovePassengers = getPassengers(passengers, currentFloor, move);
    if (currentMovePassengers.length === 0) {
      // Reverse the elevator move
      const reverseMove = getReverseMove(move);
      // Get all passengers that has opposite move with the current move of the elevator.
      currentMovePassengers = getPassengers(
        passengers,
        currentFloor,
        reverseMove,
        true
      );
      if (currentMovePassengers.length === 0) {
        // Get all passengers that has same move with the current move of the elevator where the passengers floor is less than or equal to the elevator's current floor.
        currentMovePassengers = getPassengers(
          passengers,
          currentFloor,
          move,
          true
        );
        // Reverse the elevator's move to pick up the passengers on the previous floor.
        if (currentMovePassengers.length > 0) {
          move = reverseMove;
        }
      }
      isPriorityFrom = true;
      lastFloor = getLastFloor(currentMovePassengers, reverseMove);
      maxFloor = lastFloor;
    }
    if (currentMovePassengers.length > 0) {
      // Floor that has passengers to in or out from the elevator
      const currentFloorPassengers = getCurrentFloorPassengers(
        currentMovePassengers,
        currentFloor,
        isPriorityFrom
      );
      // The elevator will be open within 10 seconds
      if (currentFloorPassengers.length > 0) {
        dispatch(changeStatus({ id, status: "open" }));
        dispatch(setInsidePassengers({ id, currentFloor }));
        if (getUnsettledFloorTo(currentFloorPassengers).length > 0) {
          setIsShowFloorForm(true);
          setIsFloorFrom(false);
        } else {
          setIsFloorFrom(true);
        }
        elevatorTime = openFloorTime;
      }
      elevatorTimeout = setTimeout(() => {
        // Passengers to be debarked
        const passengersToDebark = getPassengersToDebark(
          currentFloorPassengers,
          currentFloor,
          move
        );
        // The elevator still have passengers
        if (currentMovePassengers.length > passengersToDebark.length) {
          currentFloor !== lastFloor &&
            dispatch(
              changeStatus({
                id,
                status: move === "up" ? "Moving Up" : "Moving Down",
              })
            );
          elevatorTimeout = setTimeout(() => {
            // Elevator will start move
            if (move === "up" && currentFloor < maxFloor) {
              dispatch(moveUp({ id }));
            } else if (move === "down" && currentFloor > 1) {
              dispatch(moveDown({ id }));
            } else {
              dispatch(changeMove({ id, move: getReverseMove(move) }));
              isPriorityFrom = false;
            }
          }, [moveFloorTime]);
        }
        // Debark the passengers
        if (passengersToDebark.length > 0) {
          dispatch(debarkPassengers({ id, passengersToDebark }));
        }
      }, elevatorTime);
    } else {
      // Elevator has no passenger
      dispatch(changeStatus({ id, status: "stay" }));
      setIsFloorFrom(true);
      setIsShowFloorForm(false);
    }
    return () => clearTimeout(elevatorTimeout);
  }, [currentFloor, passengers, move]);

  const handleUserMove = (move) => {
    if (move === "up") {
      setIsFloorFrom(true);
      setIsShowFloorForm(true);
      setUserMove("up");
    } else {
      setIsFloorFrom(true);
      setIsShowFloorForm(true);
      setUserMove("down");
    }
  };

  const handleRandomCall = () => {
    const randomFloorFrom = generateRandomFloor(1, maxFloor);
    dispatch(
      requestIn({
        id,
        floorFrom: generateRandomFloor(1, maxFloor),
        floorTo: generateRandomFloor(1, maxFloor, randomFloorFrom),
        isDefaultFloorTo: false,
      })
    );
  };

  return (
    <div className="Elevator" onClick={() => null}>
      <div className="Elevator-move-requests">
        <ul>
          {passengers.map((getOutSidePassenger, index) => (
            <li key={index}>{`"${getPassengerMove(
              getOutSidePassenger.floorFrom,
              getOutSidePassenger.floorTo
            )}" request on floor ${
              getOutSidePassenger.floorFrom
            } received`}</li>
          ))}
        </ul>
      </div>
      <div className="Elevator-car">
        <div className="Elevator-name">{name}</div>
        <div className="Elevator-status">{status}</div>
        <div className="Elevator-current-floor">{currentFloor}</div>
      </div>
      <div className="Elevator-move">
        {isShowFloorForm ? (
          <FloorForm
            elevatorId={id}
            currentFloor={currentFloor}
            maxFloor={maxFloor}
            minFloor={1}
            isFloorFrom={isFloorFrom}
            userMove={userMove}
            onSubmit={() => setIsShowFloorForm(false)}
          />
        ) : (
          <>
            <button onClick={() => handleUserMove("up")}>UP</button>
            <button onClick={() => handleUserMove("down")}>DOWN</button>
            <button onClick={() => handleRandomCall()}>Random passenger</button>
          </>
        )}
      </div>
    </div>
  );
};

export default Elevator;
