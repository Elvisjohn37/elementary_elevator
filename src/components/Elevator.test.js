import { render, screen, within } from "@testing-library/react";
import Elevator from "./Elevator";
import { Provider } from "react-redux";
import store from "./../redux/store";

describe("<Elevator />", () => {
  let defaultProps = {
    currentFloor: 1,
    name: "elevator1",
    status: "stay",
    id: "1",
    maxFloor: "10",
    passengers: [],
    move: "up",
  };

  it("Should render snapshot", () => {
    let { asFragment } = render(
      <Provider store={store}>
        <Elevator {...defaultProps} />
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("Should have name", () => {
    render(
      <Provider store={store}>
        <Elevator {...defaultProps} />
      </Provider>
    );
    expect(screen.getByText(defaultProps.name)).toBeInTheDocument();
  });

  it("Should have floor indicator", () => {
    render(
      <Provider store={store}>
        <Elevator {...defaultProps} />
      </Provider>
    );
    expect(screen.getByText(defaultProps.currentFloor)).toBeInTheDocument();
  });

  it("Should start move when containing passengers", () => {
    defaultProps = {
      ...defaultProps,
      passengers: [{ floorFrom: 1, floorTo: 10 }],
    };
    render(
      <Provider store={store}>
        <Elevator {...defaultProps} />
      </Provider>
    );
    expect(screen.getByText(defaultProps.currentFloor)).toBeInTheDocument();
  });
});
