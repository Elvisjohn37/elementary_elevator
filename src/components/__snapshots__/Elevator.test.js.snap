// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Elevator /> Should render snapshot 1`] = `
<DocumentFragment>
  <div
    class="Elevator"
  >
    <div
      class="Elevator-move-requests"
    >
      <ul />
    </div>
    <div
      class="Elevator-car"
    >
      <div
        class="Elevator-name"
      >
        elevator1
      </div>
      <div
        class="Elevator-status"
      >
        stay
      </div>
      <div
        class="Elevator-current-floor"
      >
        1
      </div>
    </div>
    <div
      class="Elevator-move"
    >
      <button>
        UP
      </button>
      <button>
        DOWN
      </button>
      <button>
        Random passenger
      </button>
    </div>
  </div>
</DocumentFragment>
`;
